### PR TITLE
Itests

### LINKED ISSUE #41: Add integration tests for `ingress` and `ingress-per-unit`
We should have some basic tests in place that deploy traefik, and have it related on both the ingress, and ingress-per-unit interface.

More advanced tests should follow that switch the subdomain/path mode and test the available routes.

### LINKED ISSUE #43: When downscaling a related consumer app, `IngressPerUnitRequirer` does not sync endpoints.
### Bug Description

Let's say we deploy one `traefik-k8s` unit and three `loki-k8s` units and relate both apps. traefik charm will generate 3 endpoints and will send them over relation data.

In relation data we see those 3 endpoints:
```shell
$ juju show-unit loki/0
loki/0:
...
  - endpoint: ingress-per-unit
    related-endpoint: ingress-per-unit
    application-data:
      _supported_versions: '- v1'
      data: |-
        ingress:
          loki/0:
            url: http://192.168.122.10:80/cos-loki-0
          loki/1:
            url: http://192.168.122.10:80/cos-loki-1
          loki/2:
            url: http://192.168.122.10:80/cos-loki-2
    related-units:
      traefik-k8s/0:
        in-scope: true
        data:
          egress-subnets: 192.168.122.10/32
          ingress-address: 192.168.122.10
          private-address: 192.168.122.10
  provider-id: loki-0
  address: 10.1.157.69
```

But if we downscale loki to 1 unit: `juju scale-application loki 1` we will still have 3 endpoints:

```shell
$ juju show-unit loki/0
loki/0:
...
  - endpoint: ingress-per-unit
    related-endpoint: ingress-per-unit
    application-data:
      _supported_versions: '- v1'
      data: |-
        ingress:
          loki/0:
            url: http://192.168.122.10:80/cos-loki-0
          loki/1:
            url: http://192.168.122.10:80/cos-loki-1
          loki/2:
            url: http://192.168.122.10:80/cos-loki-2
    related-units:
      traefik-k8s/0:
        in-scope: true
        data:
          egress-subnets: 192.168.122.10/32
          ingress-address: 192.168.122.10
          private-address: 192.168.122.10
  provider-id: loki-0
  address: 10.1.157.69
```

And what is going to happen if we **upscale** loki to 4 units? will we still have 3 endpoints in relation data?
No, upscaling works as expected and we have 4 endpoints:

```shell
$ juju show-unit loki/0
loki/0:
...
  - endpoint: ingress-per-unit
    related-endpoint: ingress-per-unit
    application-data:
      _supported_versions: '- v1'
      data: |-
        ingress:
          loki/0:
            url: http://192.168.122.10:80/cos-loki-0
          loki/1:
            url: http://192.168.122.10:80/cos-loki-1
          loki/2:
            url: http://192.168.122.10:80/cos-loki-2
          loki/3:
            url: http://192.168.122.10:80/cos-loki-3
    related-units:
      traefik-k8s/0:
        in-scope: true
        data:
          egress-subnets: 192.168.122.10/32
          ingress-address: 192.168.122.10
          private-address: 192.168.122.10
  provider-id: loki-0
  address: 10.1.157.69
```

### To Reproduce

1. Deploy traefik
2. Deploy 3 units of a charm that uses `IngressPerUnitRequirer`
3. Relate both charms
4. Check you have 3 endpoints in relation data
5. Downscale consumer charm to, let's say, 2 units.
6. 4. Check you still have 3 endpoints in relation data.

### Environment

EVERYWHERE

### Relevant log output

```shell
--
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by PietroPasotti**
## Feature
adding integration tests for traefik-k8s

## Release Notes
* Added integration tests for traefik-k8s
* Fixed bug where ingress-per-unit relation data for departed units would linger after a unit had departed (and added itest for it)

Fixes #43 
Fixes #41

### CODE REVIEW

File: tests/integration/test_charm_ipa.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    }
+
+    assert_app_databag_equals(
+        info, remote_unit, expected_requirer_app_data, INGRESS_PROVIDES_APP_SCHEMA
+    )
+
+
+# cleanup before closing this test module: unrelate applications, scale requirer
+# mock back down to 1 and check final status
+async def test_reset_to_initial_state(ops_test):
```
Comment (jnsgruk) [inline tests/integration/test_charm_ipa.py:158]: Could do with a test that does relate, unrelate then relate again and check things are consistent?

**Review by jnsgruk — APPROVED** [review]
Overall LGTM. Thanks.

File: tests/integration/test_charm_ipa.py (hunk 1 above)
Comment (PietroPasotti) [reply tests/integration/test_charm_ipa.py:184]: done!

File: tests/integration/test_charm_ipu.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        }
+    }
+
+    assert_app_databag_equals(
+        info, remote_unit, expected_requirer_app_data, INGRESS_PROVIDES_APP_SCHEMA
+    )
+
+
+# cleanup before closing this test module: unrelate applications and check final status
+async def test_reset_to_initial_state(ops_test):
```
[the remark this answers is not in the document]
Comment (PietroPasotti) [reply tests/integration/test_charm_ipu.py:182]: done!

**Pull request merged**